src/lakefs_spec/spec.py

```diff
@@ -406,6 +406,57 @@ def info(self, path: str | os.PathLike[str], **kwargs: Any) -> dict[str, Any]:
             "type": "directory",
         }
 
+    def _update_dircache(self, info: list) -> None:
+        """Update logic for dircache (optionally recursive) based on lakeFS API response"""
+        parents = {self._parent(i["name"]) for i in info}
+        for pp in parents:
+            # subset of info entries which are direct descendants of `parent`
+            dir_info = [i for i in info if self._parent(i["name"].rstrip("/")) == pp]
+            if pp not in self.dircache:
+                self.dircache[pp] = dir_info
+                continue
+
+            # Merge existing dircache entry with updated listing, which contains either:
+            # - files not present in the cache yet
+            # - a fresh listing (if `refresh=True`)
+
+            cache_entry = self.dircache[pp][:]
+
+            old_names = {e["name"] for e in cache_entry}
+            new_names = {e["name"] for e in dir_info}
+
+            to_remove = old_names - new_names
+            to_update = old_names.intersection(new_names)
+
+            # Remove all entries no longer present in the current listing
+            cache_entry = [e for e in cache_entry if e["name"] not in to_remove]
+
+            # Overwrite existing entries in the cache with its updated values
+            for name in to_update:
+                old_idx = next(idx for idx, e in enumerate(cache_entry) if e["name"] == name)
+                new_entry = next(e for e in info if e["name"] == name)
+
+                cache_entry[old_idx] = new_entry
+                dir_info.remove(new_entry)
+
+            # Add the remaining (new) entries to the cache
+            cache_entry.extend(dir_info)
+            self.dircache[pp] = sorted(cache_entry, key=operator.itemgetter("name"))
+
+    def _ls_from_cache(self, path: str, recursive: bool = False) -> Any | list | None:
```

Comment:
Return type should be `list[dict[str, Any]] | None`.

Reply:
Addressed in 926fb3be2d1eee23b75dfc9514e8570acff35519.